geode-common/src/main/java/org/apache/geode/internal/Retry.java

```diff
@@ -47,45 +47,44 @@ public void sleep(long sleepTime, TimeUnit sleepTimeUnit) throws InterruptedExce
     }
   }
 
-  private static final SteadyClock steadyClock = new SteadyClock();
+  private static final SteadyTimer steadyClock = new SteadyTimer();
 
   /**
    * Try the supplier function until the predicate is true or timeout occurs.
    *
    * @param timeout to retry for
+   * @param timeoutUnit the unit for timeout
    * @param interval time between each try
-   * @param timeUnit to retry for
+   * @param intervalUnit the unit for interval
    * @param supplier to execute until predicate is true or times out
    * @param predicate to test for retry
    * @param <T> type of return value
    * @return value from supplier after it passes predicate or times out.
    */
-  public static <T> T tryFor(long timeout,
-      long interval,
-      TimeUnit timeUnit,
+  public static <T> T tryFor(long timeout, TimeUnit timeoutUnit,
+      long interval, TimeUnit intervalUnit,
       Supplier<T> supplier,
       Predicate<T> predicate) throws TimeoutException, InterruptedException {
-    return tryFor(timeout, interval, timeUnit, supplier, predicate, steadyClock);
+    return tryFor(timeout, timeoutUnit, interval, intervalUnit, supplier, predicate, steadyClock);
   }
 
   @VisibleForTesting
-  static <T> T tryFor(long timeout,
-      long interval,
-      TimeUnit timeUnit,
+  static <T> T tryFor(long timeout, TimeUnit timeoutUnit,
+      long interval, TimeUnit intervalUnit,
       Supplier<T> supplier,
       Predicate<T> predicate,
-      Clock clock) throws TimeoutException, InterruptedException {
-    long until = clock.nanoTime() + NANOSECONDS.convert(timeout, timeUnit);
+      Timer timer) throws TimeoutException, InterruptedException {
+    long until = timer.nanoTime() + NANOSECONDS.convert(timeout, timeoutUnit);
 
     T value;
     do {
       value = supplier.get();
       if (predicate.test(value)) {
         return value;
       } else {
-        clock.sleep(interval, timeUnit);
+        timer.sleep(interval, intervalUnit);
```

Comment:
The sleep should happen for min(interval, until - now) otherwise you could sleep past the timeout.

Reply:
Yes, but this extra calculation only affects the last iteration, and either we sleep past the timeout or exactly at the timeout, we would exit the loop anyway. I wouldn't mind sleeping for some extra time in the last iteration if we can make the code simpler.

Reply:
Also, we sleep at the end only to delay the return of the failed predicate state. Sleeping min time and then continue after guards against sleeping past the timeout and gives the opportunity of the predicate to go pass before returning. No point in sleeping if we are just going to fall through the while block immediately, we aren't trying to delay the return but rather the execution of the code block.

Reply:
I am not sure I follow here. if we sleep past the timeout then we won't execute the predicate anymore, it would just end the while loop with TimeoutException.

Reply:
The last iteration is the problem, why sleep to delay the return. Either just return because it would timeout at its regular interval or reduce the interval and try again before timing out. Either way there is no point I delaying the results in the last iteration as this implementation does.

Reply:
what I am saying is that even it reduces the interval it wouldn't try again, it will always just throw TimeoutException in the end.

Reply:
The example @jinmeiliao gave covers use-cases where the number of iterations is potentially large. Those are cases where time to run predicate + time to run supplier + sleep _interval_ << _timeout_.

There are other use-cases where that relation does not hold. In those cases perhaps we will only execute the body of the loop once. Imagine time to run predicate + time to run supplier + sleep _interval_ === _timeout_ i.e. the timeout interval is roughly of the same magnitude as the other work to be done each time through the loop. In that case, @pivotal-jbarrett's concern about overshooting the timeout seem more important. This second scenario seems to me to be worth the effort of the added complexity of calculating `sleepTime = min(interval, until - now)` and returning immediately if that's less-than-or-equal-to zero.

Reply:
@Bill Your example is a special case where the LAST iteration is the ONLY iteration. The extra computing only makes a difference in the LAST iteration and only when it throws a TimeoutException.

I will try to make my case one more time, if after this, you guys still insist, I will rest my case.

1. Without the extra calculation, we would NEVER delay a normal return, we will only delay a TimeoutException by a little bit (maximum the rest period). Since we are going to timeout anyway, I don't really care about this little extra wait time.
2. The extra calculation only make a difference in the last iteration when it's about to reach a timeout, for all the previous loops, it's a waste of time and an extra call to the timer.getTime().
3. `until - now` could be negative, so the calculation should really be something like `min(interval, (until - now) < 0 ? 0 : (until - now)` which completely take the beauty and readability out of the code. To me, the detriment overweights the benefit. 

But if you guys still think otherwise, do let me know. I will make the change. @Bill @pivotal-jbarrett

Reply:
I don't think any extra call is needed to implement the proposal @jinmeiliao. The call to `timer.nanoTime()` can be done in the body of the loop and used in the body and in the `while()` condition I think. I still feel like the added code complexity is worth it since the whole point of making a stand-alone `Retry` utility like this is to be able to use it in many situations.

Reply:
No, we can't reuse `timer.nanoTime()` since it will be different time at the time of calculation.

I am still not convinced, but I will go ahead and make the change and let you guys decide.

Reply:
would something like this work:

![image](https://user-images.githubusercontent.com/4002/99710928-0129c200-2a56-11eb-84fd-a1f66c757363.png)

Reply:
I think will run the supplier and predicate one more time than the previous iteration. 

Using my previous example of 10 seconds timeout and 3 seconds sleep (supplier and predicate execution time 0 and predicate always returns false), I am supposed to throw the TimeoutException after sleeping for the last 1 second , but your implementation will run supplier/predicate one more time and then throw.

Reply:
Agreed. With this approach we never return directly after a sleep. (That's what @pivotal-jbarrett meant above when he said "No point in sleeping if we are just going to fall through the while block immediately, we aren't trying to delay the return but rather the execution of the code block") We always return immediately after testing the predicate. It is different, but I don't think it's wrong.

If we are bothering to sleep, it makes sense to give the predicate one last try before returning. Otherwise why did we sleep?

Reply:
I think I get your point, we already wasted some time, why don't we give it another try? It all boils down to what you think is right or tolerable.

Reply:
you know if you change your implementation a little bit, instead of `if (sleepNanos > 0)` we change it to `if(sleepNanos == intervalNanos)`, I think this would be best.

Reply:
Ok, I think my last commit combines the good points of both implementation. Thanks for all the points and nudges.